packages/catlog/src/stdlib/analyses/ecld/atomisations.rs

```diff
@@ -0,0 +1,292 @@
+/*! Atomisations for extended causal loop diagrams (ECLDs)
+
+ECLDs have arrows labelled by two natural numbers, for degree and delay. In the
+intended linear ODE semantics, both of these behave additively under composition
+of paths. It is useful to have a rewrite rule that "atomises" any arrow, i.e.
+replacing an arrow X -> Y of degree n (which corresponds to the equation
+(d/dt)^n(Y) += kX) by n-many arrows of degree 1, thus also introducing (n-1)-many
+new objects X -> Y_{n-1} -> ... -> Y_2 -> Y_1 -> Y. The idea to keep in mind is
+that a degree-n differential equation of the form (d/dt)^n(Y) = X can
+equivalently be written as a system of degree-1 differential equations, namely
+(d/dt)(Y) = Y_1, (d/dt)(Y_1) = Y_2, ..., (d/dt)(Y_{n-1}) = X. An analogous story
+holds for order of delay, though this is formally dual: an arrow X -> Y of
+order m corresponds to (something like) a morphism Y += k*E(m)*X, i.e. the
+*source* is modified, not the target.
+
+We call the objects Y_i the *formal derivatives* (resp. *formal delays*) of Y,
+and the list [Y_0 = Y, Y_1, ..., Y_{n-1}] the *tower* over Y (resp. under Y),
+and call Y the *base* of this tower; we call the length n of the list
+[Y_0, ..., Y_{n-1}] the *height* of the tower.
+ */
+
+use crate::dbl::model::{DiscreteDblModel, FgDblModel, MutDblModel};
+use crate::one::{Path, category::FgCategory};
+use crate::stdlib::theories;
+use crate::zero::{QualifiedName, name};
+use std::{collections::HashMap, rc::Rc};
+
+// Some helpful functions
+fn deg_of_mor(model: &DiscreteDblModel, f: &QualifiedName) -> usize {
+    model.mor_generator_type(f).into_iter().filter(|t| *t == name("Degree")).count()
+}
+
+fn is_mor_pos(model: &DiscreteDblModel, f: &QualifiedName) -> bool {
+    0 == model
+        .mor_generator_type(f)
+        .into_iter()
+        .filter(|t| *t == name("Negative"))
+        .count()
+        % 2
+}
```

Comment:
I feel like we ought to be doing such reasoning using general machinery rather than ad hoc calculations. However, the procedure that comes to mind, namely

1. Project onto the monoid of signs
2. Check whether the result is equal (in the e-graph) to the generator Negative

is almost certainly more code, so perhaps not worth the additional complexity.

Reply:
Come to think of it, to enable the expected model migrations, we're going to have to define the needed theory map anyway, similar to [this one](https://github.com/ToposInstitute/CatColab/blob/49b3f4d9244d95e5388a14f28314e987d6d47161/packages/catlog/src/stdlib/theory_morphisms.rs#L40), so maybe it *is* worth it after all.